staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_source.go

```diff
@@ -0,0 +1,458 @@
+	// with it.
+	if genericInformer, err := s.informerFactory.ForResource(mapping.Resource); err == nil {
+		informer = genericInformer
+
+		// Start the informer
+		s.informerFactory.Start(instanceContext.Done())
+
+	} else {
+		// Dynamic JSON informer fallback.
+		// Cannot use shared dynamic informer since it would be impossible
+		// to clean CRD informers properly with multiple dependents
+		// (cannot start ahead of time, and cannot track dependencies via stopCh)
+		informer = dynamicinformer.NewFilteredDynamicInformer(
+			s.dynamicClient,
+			mapping.Resource,
+			corev1.NamespaceAll,
+			// Use same interval as is used for k8s typed sharedInformerFactory
+			// https://github.com/kubernetes/kubernetes/blob/7e0923899fed622efbc8679cca6b000d43633e38/cmd/kube-apiserver/app/server.go#L430
+			10*time.Minute,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+			nil,
+		)
+		go informer.Informer().Run(instanceContext.Done())
+	}
+
+	controller := generic.NewController(
+		generic.NewInformer[runtime.Object](informer.Informer()),
+		func(namepace, name string, newObj runtime.Object) error {
+			// do nothing
+			return nil
+		},
+		generic.ControllerOptions{
+			Workers: 1,
+			Name:    paramSource.String() + "-controller",
+		},
+	)
+
+	ret := &paramInfo{
+		mapping:    *mapping,
+		cancelFunc: instanceCancel,
+		informer:   informer,
+	}
+	s.paramsCRDControllers[*paramSource] = ret
+	go func() {
+		stopReason := controller.Run(instanceContext)
+		if stopReason != nil && !goerrors.Is(stopReason, context.Canceled) {
+			utilruntime.HandleError(fmt.Errorf("param controller %v stopped unexpectedly: %v", paramSource, stopReason))
+		}
+	}()
+
+	return ret.informer, mapping.Scope, nil
+}
+
+// For testing
+func (s *policySource[P, B]) getParamInformer(param schema.GroupVersionKind) (informers.GenericInformer, meta.RESTScope) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if info, ok := s.paramsCRDControllers[param]; ok {
+		return info.informer, info.mapping.Scope
+	}
+
+	return nil, nil
+}
+
+func (s *policySource[P, B]) compilePolicyLocked(policySpec P) Evaluator {
+	policyMeta, err := meta.Accessor(policySpec)
+	if err != nil {
+		// This should not happen if P, and B have ObjectMeta, but
+		// unfortunately there is no way to express "able to call
+		// meta.Accessor" as a type constraint
+		utilruntime.HandleError(err)
+		return nil
+	}
+	key := types.NamespacedName{
+		Namespace: policyMeta.GetNamespace(),
+		Name:      policyMeta.GetName(),
+	}
+
+	compiledPolicy, wasCompiled := s.compiledPolicies[key]
+
+	// If the policy or binding has changed since it was last compiled,
+	// and if there is no configuration error (like a missing param CRD)
+	// then we recompile
+	//!TODO: also recompile if ParamCRD changed???
+	// Don't think so, typechecking is done in kube-controller manager,
+	// Compiled expressions seem to use lazy field validation?
+	// should ask @cici37 to clarify
```

Comment:
Ideally we should also watch for CRD changes but there is existing concerns on this. I'll schedule a followup meeting on this one :)

Reply:
after our followup we decided waiting for the restmapper was sufficient. We can investigate ways go make it better on startup (like manually resetting it after initial CRDs are visible in discovery)